website/blog/2024-12-16-custom-tracing-provider/index.md

```diff
@@ -0,0 +1,376 @@
+2. We define a decorator, `trace_groq_chat`, using [`functools.wraps`](https://docs.python.org/3/library/functools.html#functools.wraps), that wraps the `chat.completions.create` method. There are a few key steps here:
+    1. We start a new span with `mlflow.start_span`. The span name is set to the fully qualified name of the method, which is available via `func.__qualname__`. The span type is set to the value returned by `_get_span_type`. Note that we might consider alternate approaches for automatically determining the span type. For example, for a client that had both `embeddings.create` and `chat.completions.create` methods, we would be more interested in the "embeddings" and "chat" parts of the name, respectively.
+    2. We set `model_name` as an attribute on the span with `span.set_attribute`. This isn't strictly necessary as model name will be captured in the inputs, but it illustrates how to set arbitrary attributes on a span.
+    3. We log the messages as inputs to the span with `span.set_inputs`. We get these from the `messages` argument to the `chat.completions.create` method by accessing the `kwargs` dictionary. These messages will be logged to the "inputs" section of the span in the MLflow UI. We also log the model name as an input, again to illustrate how to record arbitrary inputs.
+
+        ![Logging inputs](./2_log_inputs.png)
+
+    4. We call the original method with `func(*args, **kwargs)`. This is the Groq `chat.completions.create` method.
+    5. We log the outputs of the method as a span attribute with `span.set_outputs`. This takes the dictionary returned by the Groq API and sets it as an attribute on the span. This outputs will be logged to the "outputs" section of the span in the MLflow UI. If we wanted, we could log only a subset of the outputs, or log them in a different format. But in many case, especially for LLM providers that return results in a format similar to the OpenAI spec, simply logging all of the output works well and results in informative and readable traces.
+
+        ![Logging outputs](./3_log_outputs.png)
+
+    6. Finally, we return the response from the API call, without any changes. Now, when we wrap a client instance with `trace_groq_chat`, the `chat.completions.create` method will be traced, but will otherwise behave as normal.
+
+A few points to note:
+
+- We are wrapping a method on a client *instance*, not a class. This is a fairly lightweight approach that does what we need it to do without requiring changes to the Groq SDK code.
```

Comment:
Might want to use the term "patching" instead of wrapping. We are still guerilla patching here by overriding the existing implementation. Wrapping would be more of using the instance of the return obj + method reference of the return value of trace_groq_chat.